module/zfs/vdev_initialize.c

```diff
@@ -423,10 +423,24 @@ vdev_initialize_ms_mark(metaslab_t *msp)
 }
 
 static void
-vdev_initialize_ms_unmark(metaslab_t *msp)
+vdev_initialize_ms_unmark(spa_t *spa, metaslab_t *msp)
 {
 	ASSERT(!MUTEX_HELD(&msp->ms_lock));
 	metaslab_group_t *mg = msp->ms_group;
+
+	/*
+	 * When the last thread has finished initializing a metaslab and
+	 * new allocations are about to be allowed.  Then it's critical
+	 * to wait for any initializing writes to be synced or we risk
+	 * overwriting newly allocated blocks.  It is harmless for another
+	 * another thread to remark this metaslab while waiting on the sync.
+	 */
+	mutex_enter(&msp->ms_lock);
+	boolean_t sync = (msp->ms_initializing == 1);
+	mutex_exit(&msp->ms_lock);
+	if (sync)
+		txg_wait_synced(spa->spa_dsl_pool, 0);
+
```

Comment:
This still leaves a potential race since we could have 2 threads checking the `ms_initializing` flag and finding the value does not equal 1. So both don't call `txg_wait_sync` but later decrement the count to 0.

To solve this we need to first grab the `mg_ms_initialize_lock`. Here's a proposed restructure:
```
        spa_t *spa = mg->mg_vdev->vd_spa;

        mutex_enter(&mg->mg_ms_initialize_lock);
        mutex_enter(&msp->ms_lock);
        if (--msp->ms_initializing == 0) {
                mutex_exit(&msp->ms_lock);
                mutex_exit(&mg->mg_ms_initliaze_lock);

                /* Drop the locks before syncing out the txg */
                txg_wait_synced(spa_get_dsl(spa), 0);
                mutex_enter(&mg->mg_ms_initialize_lock);
                mutex_enter(&msp->ms_lock);
                cv_broadcast(&mg->mg_ms_initialize_cv);
        }
        mutex_exit(&msp->ms_lock);
        mutex_exit(&mg->mg_ms_initialize_lock);
```
I think this will work. :) This should reduce the number of times we have to call `txg_wait_synced`.

Reply:
@grwilson you're right about the race in the current version.  However, I don't believe this version is suitable either.  My concern is that the `txg_wait_synced()` needs to happen before `ms_initializing == 0` otherwise new allocations will be allowed too soon.  I'm inclined to go back to the simpler version for now even if it means extra calls to `txg_wait_synced`.

Reply:
Ah yes, this should do it:

```
        spa_t *spa = mg->mg_vdev->vd_spa;

        mutex_enter(&mg->mg_ms_initialize_lock);
        mutex_enter(&msp->ms_lock);
        sync == (msp->ms_initializing == 1);
        if (sync) {
                mutex_exit(&msp->ms_lock);
                txg_wait_synced(spa_get_dsl(spa), 0);
                mutex_enter(&msp->ms_lock);
        }

        if (--msp->ms_initializing == 0) {
                mg->mg_ms_initializing--;
                cv_broadcast(&mg->mg_ms_initialize_cv);
        }
        mutex_exit(&msp->ms_lock);
        mutex_exit(&mg->mg_ms_initialize_lock);```
My concern is this might serialize the initialization. So not sure if this is actually an optimization over doing a `txg_wait_synced` more than once like you had originally.

Reply:
> My concern is this might serialize the initialization. 

This was my concern as well, and it's compounded by the refactoring proposed in #8470 which I've folded in to the trim PR.  This could result in initialize, trim, and raidz expansion all serializing on each other.  Let's go with a variant of the original version, I've updated the trim PR to included the [updated version](https://github.com/zfsonlinux/zfs/pull/8419/files#diff-9fd6b453f8153161abe0728c449e6505R4731) which looks like this:

```c
void
metaslab_enable(metaslab_t *msp, boolean_t sync)
{
	metaslab_group_t *mg = msp->ms_group;
	spa_t *spa = mg->mg_vd->vdev_spa;

	/*
	 * When enabling a metaslab it may need to wait for outstanding I/Os
	 * to be synced.  The initialize and TRIM features depend on this
	 * sync to prevent newly allocated blocks from being overwritten.
	 */
	if (sync)
		txg_wait_synced(spa_get_dsl(spa), 0);

	mutex_enter(&mg->mg_ms_disabled_lock);
	mutex_enter(&msp->ms_lock);
	if (--msp->ms_disabled == 0) {
		mg->mg_ms_disabled--;
		cv_broadcast(&mg->mg_ms_disabled_cv);
	}
	mutex_exit(&msp->ms_lock);
	mutex_exit(&mg->mg_ms_disabled_lock);
}
```